src/kudu/consensus/log_cache.cc

```diff
@@ -619,7 +619,8 @@ Status LogCache::BlockingReadOps(int64_t after_op_index,
 
   {
     std::lock_guard<Mutex> l(lock_);
-    while (after_op_index >= next_sequential_op_index_) {
+
+    while ((after_op_index + 1) >= next_sequential_op_index_) {
```

Comment:
This diff triggered a thought.Is it possible that due to the wait at the proxy node,
we are increasing the probability that the destination node will timeout more often.
500ms wait time here + typical 500 ms wait on LEADER for heartbeats, could it mean
that we have stretched out potential delay on final destination ?

Reply:
why is the condition on line 630 below not changed? Does the off by 1 issue not manifest there?

Reply:
Good catch! The check needs to be in line 630. It was not causing a problem because ReadOps(...) does the same check again and returns an error anyways. Have fixed the check in 630 to be safe.

Regarding timeouts on destination, it is possible. However, heartbeat messages are not blocked - since heartbeat does not contain any ops there is no need to read from log. So, heartbeats are forwarded by the proxy immediately. We can test with prod like workload and see if we have to handle this in a different way.